flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,361 @@
+                List<SchedulingExecutionVertex> executionVertices,
+                int index) {
+            final int slotNum =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup).size();
+            final int jobVertexParallel = executionVertices.size();
+            if (slotNum == jobVertexParallel) {
+                // For CoLocations with max parallelism of the slot sharing group.
+                allocateCoLocatedMaxParallelVertices(slotSharingGroup, executionVertices, index);
+            } else {
+                // For CoLocations with non-max parallelism of the slot sharing group.
+                allocateCoLocatedNonMaxParallelVertices(slotSharingGroup, executionVertices);
+            }
+        }
+
+        private void allocateCoLocatedMaxParallelVertices(
+                SlotSharingGroup slotSharingGroup,
+                List<SchedulingExecutionVertex> executionVertices,
+                int index) {
+            final List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup);
+            for (SchedulingExecutionVertex executionVertex : executionVertices) {
+                final CoLocationConstraint coLocationConstraint =
+                        getCoLocationConstraint(executionVertex);
+                ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        constraintToExecutionSlotSharingGroupMap.get(coLocationConstraint);
+                if (Objects.isNull(executionSlotSharingGroup)) {
+                    executionSlotSharingGroup = executionSlotSharingGroups.get(index);
+                }
+                addVertexToExecutionSlotSharingGroup(executionSlotSharingGroup, executionVertex);
+                index = ++index % executionSlotSharingGroups.size();
+            }
+            updateConstraintToExecutionSlotSharingGroupMap(executionVertices);
+        }
+
+        private void allocateCoLocatedNonMaxParallelVertices(
+                SlotSharingGroup slotSharingGroup,
+                List<SchedulingExecutionVertex> executionVertices) {
+            List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup);
+            int index;
+            // For CoLocations with non-max parallelism of the slot sharing group.
+            for (SchedulingExecutionVertex executionVertex : executionVertices) {
+                final CoLocationConstraint coLocationConstraint =
+                        getCoLocationConstraint(executionVertex);
+                ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        constraintToExecutionSlotSharingGroupMap.get(coLocationConstraint);
+                if (Objects.isNull(executionSlotSharingGroup)) {
+                    index = getFittestSlotIndex(executionSlotSharingGroups, executionVertex);
```

Comment:
The time complexity here is O(n*n). I'm not sure whether it will affect the Flink job start duration.

Anyway, it's better to do a benchmark for this FLIP in the end.

Reply:
SGTM +1. 👍

Reply:
Would you mind creating an JIRA to follow it?